### PR TITLE
Support Google OAuth2 and other OIdC providers

### PR DESCRIPTION
**Pull request by EngHabu**
# TL;DR
A few changes to make sure OIdC with other providers (besides Okta) also work:

- Move OAuth scopes to a config. This allows flyte admins to choose which additional scopes to request. E.g. request offline_access in Okta case but not in Google Auth
- Rely on id_token instead of access_token for authentication
- Persist/restore id_token to cookies

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [X] Any pending items have an associated Issue

## Tracking Issue
https://github.com/lyft/flyte/issues/657

### CODE REVIEW

File: pkg/auth/handlers.go [hunk 1]
```diff
@@ -40,18 +39,15 @@ func RefreshTokensIfExists(ctx context.Context, authContext interfaces.Authentic
...
+		idToken, accessToken, refreshToken, err := authContext.CookieManager().RetrieveTokenValues(ctx, request)
+		if err == nil {
+			_, err = ParseAndValidate(ctx, authContext.Claims(), idToken, authContext.OidcProvider())
+			if err != nil && errors.IsCausedBy(err, ErrTokenExpired) && len(refreshToken) > 0 {
+				logger.Debugf(ctx, "Expired id token found, attempting to refresh")
 				newToken, e := GetRefreshedToken(ctx, authContext.OAuth2Config(), accessToken, refreshToken)
 				err = e
 				if err == nil {
-					logger.Debugf(ctx, "Access token refreshed. Saving new tokens into cookies.")
+					logger.Debugf(ctx, "Tokens're refreshed. Saving new tokens into cookies.")
```
Comment (katrogan) [inline pkg/auth/handlers.go:50]: nit: Tokens're?

**Review by katrogan — APPROVED** [review]

**Comment by katrogan** [comment]
looks like `github.com/lyft/flyteadmin/pkg/auth` compilation failed?

File: flyteadmin_config.yaml [hunk 2]
```diff
@@ -6,12 +6,16 @@ server:
   httpPort: 8088
   grpcPort: 8089
   grpcServerReflection: true
+  kube-config: "/Users/haythamabuelfutuh/.kube/config"
```
Comment (kumare3) [inline flyteadmin_config.yaml:9]: oops!

File: pkg/auth/cookie.go [hunk 3]
```diff
@@ -17,13 +17,18 @@ import (
 
 const (
 	// #nosec
-	accessTokenCookieName = "flyte_jwt"
+	accessTokenCookieName = "flyte_at"
```
Comment (kumare3) [inline pkg/auth/cookie.go:20]: is this a backwards incompatible change?

Comment (EngHabu) [reply pkg/auth/cookie.go:20]: it'll force a re-auth... which is fine... correctness FTW

File: pkg/auth/cookie_manager.go [hunk 4]
```diff
@@ -20,6 +20,8 @@ const (
 	ErrB64Decoding errors.ErrorCode = "BINARY_DECODING_FAILED"
 	// #nosec
 	ErrTokenNil errors.ErrorCode = "EMPTY_OAUTH_TOKEN"
+	// #nosec
```
Comment (katrogan) [inline pkg/auth/cookie_manager.go:23]: huh I wonder do we still need this? can't remember if it was lyft security scanning that required this originally

File: pkg/auth/cookie_manager.go [hunk 5]
```diff
@@ -76,22 +75,37 @@ func (c CookieManager) SetTokenCookies(ctx context.Context, writer http.Response
...
+	http.SetCookie(writer, &atCookie)
+
+	if idTokenRaw, converted := token.Extra(idTokenExtra).(string); converted {
+		idCookie, err := NewSecureCookie(idTokenCookieName, idTokenRaw, c.hashKey, c.blockKey)
+		if err != nil {
+			logger.Errorf(ctx, "Error generating encrypted id token cookie %s", err)
+			return err
+		}
+
+		http.SetCookie(writer, &idCookie)
```
Comment (katrogan) [inline pkg/auth/cookie_manager.go:93]: why is it now required to set an idCookie?

Comment (EngHabu) [reply pkg/auth/cookie_manager.go:94]: Per OIdC spec, it's a required response parameter: https://openid.net/specs/openid-connect-core-1_0.html#TokenResponse

So we are guaranteed to get it from any OIdC provider after a successful authN...

File: pkg/auth/cookie_manager.go (hunk 4 above)
Comment (EngHabu) [reply pkg/auth/cookie_manager.go:23]: Apparently we still do... golangci-lint complains when I remove them...

**Comment by katrogan** [comment]
LGTM but will let @wild-endeavor review & approve

**Review by wild-endeavor — APPROVED** [review]
reminder to @anandswaminathan that when this goes out you will need to add scopes

**Comment by EngHabu** [comment]
I added them as default values: https://github.com/flyteorg/flyteadmin/pull/147/files#diff-a3d824da3c42420cd5cbb0a4a2c0e7b5bfddd819652788a0596d195dc6e31fa5R55-R60 

so we don't have to break existing behavior... Is that enough?

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]
